{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Read Alignment\n",
    "\n",
    "There are no questions in this section."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Performing Read Alignment\n",
    "\n",
    "__1.__ 145,441,459\n",
    "\n",
    "__2.__ The sam is ~ 157M and the bam is ~ 25M\n",
    "\n",
    "__3.__ 67,461 (17.2%). This is the number of unpaired read dups (359) + the number of paired read dups *2 (33551 * 2)\n",
    "\n",
    "__4.__ 392,820\n",
    "\n",
    "__5.__ 391,603/392,820 (99.7%)\n",
    "\n",
    "__6.__ 389,410\n",
    "\n",
    "__7.__ 0\n",
    "\n",
    "__8.__ 419 (mean) 113.9 (standard deviation)\n",
    "\n",
    "__9.__ 7,853 (2.0%)\n",
    "\n",
    "__10.__ First/Forward read"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Alignment Visualisation\n",
    "\n",
    "__1.__ This exercise is just asking you to explore the genome and become familiar with navigating in IGV.\n",
    "\n",
    "__2.__ 23X-57X\n",
    "\n",
    "__3.__ \n",
    "\n",
    "There is a 1bp insertion (at \"T)\" at position 87,483,966. This is supported by 9 reads.\n",
    "\n",
    "There is a 28bp deletion at position 87,483,966.  This is supported by 3 reads.\n",
    "\n",
    "The third mutation is a bit harder to spot, because it's bigger than the read length (so no single read will span it). Look first at the coverage track, and notice a sharp coverage drop between chr7:87,483,833 and chr7:87,484,169. This suggests that one of the two alleles has been deleted across that position. Notice also the soft-clipping of some reads \"entering\" chr7:87,483,833 from the left, and the same softclipping of reads entering chr7:87,484,169 from the right. This soft-clipping shows up as reads being partly multi-coloured. That's happening because the physical genome between those points has been excised for one allele, causing the mis-alignment when we attempt to align some reads to the reference genome. That mis-alignment causes the bwa aligner to \"give up\" and mark a part of the read as soft-clipped.\n",
    "\n",
    "__4.__\n",
    "\n",
    "This mouse was bred from a zygote which was mutagenised with Crispr-Cas9, targeted at the Tyr locus. You are watching the zygote DNA-repair machinery panicking and grabbing at straws when trying to repair double-stranded DNA breaks. In the process, it makes mistakes, and those mistakes are propagated into the mouse genome: different zygote cells received different mutations, which is why some reads reflect different mutations to others.\n",
    "Specifically -\n",
    "The CRISPR-Cas9 has acted on the zygote at this locus to create Non-Homologous-End-Join-based damage around 87,483,960: that resulted in a subclonal 1bp insertion and a 28bp deletion. Also, a related DNA repair process has resulted in the a 336bp deletion across the same area."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Alignment Workflows\n",
    "\n",
    "__1.__ -M marks shorter split hits as secondary and -R adds the read group to the header of the BAM file\n",
    "\n",
    "__2.__ -b means create a BAM as output and -S indicates that the input file is a SAM file. The -S option is now ignored by samtools as it can autodetect the input file type.\n",
    "\n",
    "__3.__ 397,506\n",
    "\n",
    "__4.__ 303,036/397,506 (76.2%)\n",
    "\n",
    "__5.__ 282,478\n",
    "\n",
    "__6.__ 2,239\n",
    "\n",
    "__7.__ 275.9 (mean) and 47.7 (standard deviation)\n",
    "\n",
    "__8.__ 23,789 (7.9%)\n",
    "\n",
    "__9.__ First\n",
    "\n",
    "__10.__ \n",
    "\n",
    "`bwa mem -M -R \\\"\\@RG\\\\tID:lane2\\\\tSM:60A_Sc_DBVPG6044\\\" ../../../../ref/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa.gz s_7_1.fastq.gz s_7_2.fastq.gz | samtools view -bS - | samtools sort -T temp -O bam -o lane2.sorted.bam -`\n",
    "\n",
    "~22M\n",
    "\n",
    "Merge:\n",
    "`~/course_data/read_alignment/data/Exercise2/60A_Sc_DBVPG6044/library1$ picard MergeSamFiles -I lane1/lane1.sorted.bam -I lane2/lane2.sorted.bam -O library1.bam`\n",
    "\n",
    "Markdup:\n",
    "`~/course_data/read_alignment/data/Exercise2/60A_Sc_DBVPG6044/library1$ picard MarkDuplicates -I library1.bam -O library1.markdup.bam -M library1.metrics.txt`\n",
    "\n",
    "__11.__ 12399 or 3115 + (4642 * 2) = unpaired read dups + (paired read dups *2)\n",
    "\n",
    "__12.__ 2.5%\n",
    "\n",
    "## Exercises\n",
    "\n",
    "__1.__ No answer needed\n",
    "\n",
    "__2.__ The reference base is C\n",
    "\n",
    "__3.__ No (the reads call T)\n",
    "\n",
    "__4.__ The reference base is G and all reads agree\n",
    "\n",
    "__5.__ No answer\n",
    "\n",
    "__6.__ An insertion\n",
    "\n",
    "__7.__ No answer\n",
    "\n",
    "__8.__ A deletion. This is unlikely to be a true variant and may be due to misalignment due the run of T's in the flanking region.\n",
    "\n",
    "__9.__ The following command produces a CRAM file which should be ~29MB in size.\n",
    "\n",
    "`samtools view -C -T ../../../ref/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa -o library1.markdup.cram library1.markdup.bam`\n",
    "\n",
    "-C means create a CRAM file as output\n",
    "\n",
    "-T is the reference file to use for the compression\n",
    "\n",
    "-o is the name of CRAM file to create"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
